### PR TITLE
Aot gan dev

### PR DESCRIPTION
**Pull request by quincylin1**
add randomresizedcrop, randomrotation and colorjitter
adopted from torch.transform

### CODE REVIEW

**Comment by ckkelvinchan** [comment]
Hello @quincylin1, thank you for your PR. May I ask whether you are implementing functions for the paper `AOT-GAN: Aggregated Contextual Transformations for High-Resolution Image Inpainting`?

Moreover, as transforming from NumPy array to PIL requires additional processing time, it could result in slower training time. We suggest keeping them as NumPy arrays throughout the pipeline. 

For example, the rotation function can be done by sampling a degree from `[min, max]`, and then using `cv2.rotate`.

File: mmedit/models/backbones/encoder_decoders/aot_encoder_decoder.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import torch.nn as nn
+from mmcv.runner import auto_fp16, load_checkpoint
+
+from mmedit.models.builder import build_component
+from mmedit.models.registry import BACKBONES
+from mmedit.utils import get_root_logger
+
+
+@BACKBONES.register_module()
+class AOTEncoderDecoder(nn.Module):
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/aot_encoder_decoder.py:10]: It seems that this class is identical to `GLEncoderDecoder`. You can inherit from `GLEncoderDecoder` by using
```
class AOTEncoderDecoder(GLEncoderDecoder):
```

In this case, you do not need to rewrite the function here. But please make sure the docstring is modified.

File: mmedit/models/backbones/encoder_decoders/decoders/aot_decoder.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    This implementation follows:
+    Globally and locally Consistent Image Completion
+
+    Args:
+        in_channels (int): Channel number of input feature.
+        norm_cfg (dict): Config dict to build norm layer.
+        act_cfg (dict): Config dict for activation layer, "relu" by default.
+        out_act (str): Output activation type, "clip" by default. Noted that
+            in our implementation, we clip the output with range [-1, 1].
+    """
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/decoders/aot_decoder.py:22]: Modify the docstring

File: mmedit/models/backbones/encoder_decoders/decoders/aot_decoder.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        self.dec3 = ConvModule(
+            64,
+            3,
+            kernel_size=3,
+            stride=1,
+            padding=1, 
+            act_cfg=None)
+
+        self.output_act = nn.Tanh()
+        '''
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/decoders/aot_decoder.py:62]: Remove the comments

File: mmedit/models/backbones/encoder_decoders/decoders/aot_decoder.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        Returns:
+            torch.Tensor: Output tensor with shape of (n, c, h', w').
+        """
+        '''
+        for i in range(3):
+            if i <= 1:
+                x = F.interpolate(x, scale_factor=2, mode='bilinear', align_corners=True)
+            x = getattr(self, f'dec{i + 1}')(x)
+        x = self.output_act(x)
+        '''
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/decoders/aot_decoder.py:80]: Remove the comments

File: mmedit/models/backbones/encoder_decoders/decoders/aot_decoder.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        norm_cfg (dict): Config dict to build norm layer.
+        act_cfg (dict): Config dict for activation layer, "relu" by default.
+        out_act (str): Output activation type, "clip" by default. Noted that
+            in our implementation, we clip the output with range [-1, 1].
+    """
+
+    def __init__(self,
+                 in_channels=256,
+                 act_cfg=dict(type='ReLU'),
+                 out_act='clip'):
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/decoders/aot_decoder.py:27]: `act_cfg` and `out_act` are not used.

File: mmedit/models/backbones/encoder_decoders/decoders/aot_decoder.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            x = getattr(self, f'dec{i + 1}')(x)
+        x = self.output_act(x)
+        '''
+        x = self.decoder(x)
+        x = torch.tanh(x)
+        
+        return x
+
+class UpConv(nn.Module):
+    def __init__(self, inc, outc, scale=2):
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/decoders/aot_decoder.py:87]: Please use `in_channels` instead of `inc` and `out_channels` instead of `outc`.

Also, `scale` is not used.

File: mmedit/models/backbones/encoder_decoders/encoders/aot_encoder.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            nn.ReflectionPad2d(3),
+            nn.Conv2d(4, 64, 7),
+            nn.ReLU(True),
+            nn.Conv2d(64, 128, 4, stride=2, padding=1),
+            nn.ReLU(True),
+            nn.Conv2d(128, 256, 4, stride=2, padding=1),
+            nn.ReLU(True)
+        )
+        
+        '''
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/encoders/aot_encoder.py:36]: Remove the comments

File: mmedit/models/backbones/encoder_decoders/encoders/aot_encoder.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+            
+        self.enc3 = ConvModule(
+            128, 
+            256,
+            kernel_size=4,
+            stride=2,
+            padding=1,
+            act_cfg=act_cfg)
+        
+        # self.init_weights()
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/encoders/aot_encoder.py:62]: Remove the comments

File: mmedit/models/backbones/encoder_decoders/encoders/aot_encoder.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+                if hasattr(m, 'bias') and m.bias is not None:
+                    nn.init.constant_(m.bias.data, 0.0)
+
+        self.apply(init_func)
+
+        # propagate to children
+        for m in self.children():
+            if hasattr(m, 'init_weights'):
+                m.init_weights(init_type, gain)
+    """
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/encoders/aot_encoder.py:102]: Remove

File: mmedit/models/backbones/encoder_decoders/encoders/aot_encoder.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        
+
+    def forward(self, x):
+        """Forward Function.
+
+        Args:
+            x (torch.Tensor): Input tensor with shape of (n, c, h, w).
+
+        Returns:
+            torch.Tensor: Output tensor with shape of (n, c, h', w').
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/encoders/aot_encoder.py:112]: What are `h'` and `w'`?

File: mmedit/models/backbones/encoder_decoders/encoders/aot_encoder.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        """
+
+        
+        x = self.reflectionpad2d(x)
+        for i in range(3):
+            x = getattr(self, f'enc{i + 1}')(x)
+        '''
+        x = self.encoder(x)
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/encoders/aot_encoder.py:120]: Remove

File: mmedit/models/backbones/encoder_decoders/necks/aot_neck.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+    Args:
+        in_channels (int): Channel number of input feature.
+        conv_type (str): The type of conv module. In DeepFillv1 model, the
+            `conv_type` should be 'conv'. In DeepFillv2 model, the `conv_type`
+            should be 'gated_conv'.
+        norm_cfg (dict): Config dict to build norm layer.
+        act_cfg (dict): Config dict for activation layer, "relu" by default.
+        kwargs (keyword arguments).
+    """
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/necks/aot_neck.py:23]: Modify docstring

File: mmedit/models/backbones/encoder_decoders/necks/aot_neck.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+    def __init__(self,
+                 in_channels=256,
+                 act_cfg=dict(type='ReLU'),
+                 dilation_rates='1+2+4+8',
+                 num_aotblock=8,
+                 **kwargs):
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/necks/aot_neck.py:30]: Some arguments are not used.

File: mmedit/models/backbones/encoder_decoders/necks/aot_neck.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+        for i in range(self.num_aotblock):
+            self.model.add_module(
+                f"aotblock{i + 1}", 
+                AOTBlock(
+                    in_channels=in_channels,
+                    dilation_rates=self.dilation_rates,
+                    act_cfg=act_cfg,
+                )
+            )
+        '''
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/necks/aot_neck.py:53]: Remove

File: mmedit/models/backbones/encoder_decoders/necks/aot_neck.py [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+        
+        
+        
+    def forward(self, x):
+        x = self.middle(x)
+        return x
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/necks/aot_neck.py:59]: use `return self.middle(x)` directly

File: mmedit/models/backbones/encoder_decoders/necks/aot_neck.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+
+
+class AOTBlock(nn.Module):
+    def __init__(self, dim, rates):
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/necks/aot_neck.py:63]: docstring

File: mmedit/models/backbones/encoder_decoders/necks/aot_neck.py [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+    def forward(self, x):
+        out = [self.__getattr__(f'block{str(i).zfill(2)}')(x) for i in range(len(self.rates))]
+        out = torch.cat(out, 1)
+        out = self.fuse(out)
+        mask = my_layer_norm(self.gate(x))
+        mask = torch.sigmoid(mask)
+        return x * (1 - mask) + out * mask
+
+
+def my_layer_norm(feat):
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/necks/aot_neck.py:89]: Is this a special operation, different from the conventional layer normalization?

File: mmedit/models/backbones/encoder_decoders/necks/aot_neck.py [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+        return x
+
+    def forward(self, x):
+        out = [self.__getattr__(f'block{i + 1}')(x) for i in range(len(self.rates))]
+        out = torch.cat(out, 1)
+        out = self.fuse(out)
+        mask = self.norm(self.gate(x))
+        mask = torch.sigmoid(mask)
+        return x * (1 - mask) + out * mask
+'''
```
Comment (ckkelvinchan) [inline mmedit/models/backbones/encoder_decoders/necks/aot_neck.py:188]: Remove

**Review by ckkelvinchan — COMMENTED** [review]
Hi @quincylin1, here are some comments. I will go through the others soon

**Comment by quincylin1** [comment]
> Hello @quincylin1, thank you for your PR. May I ask whether you are implementing functions for the paper `AOT-GAN: Aggregated Contextual Transformations for High-Resolution Image Inpainting`?
> 
> Moreover, as transforming from NumPy array to PIL requires additional processing time, it could result in slower training time. We suggest keeping them as NumPy arrays throughout the pipeline.
> 
> For example, the rotation function can be done by sampling a degree from `[min, max]`, and then using `cv2.rotate`.

**Comment by quincylin1** [comment]
@ckkelvinchan thank you for your comments! Sorry for the late reply I have been busy recently. Indeed this is the reimplementation of the most recent image inpainting paper "Aggregated Contextual Transformations for High-Resolution Image Inpainting".  The aim is to integrate AOT-GAN into mmediting.

I have modified the data augmentation functions so that they only use numpy arrays throughout (please see commit 327bc7f for details).

I have also pushed the most recent code for the model and evaluation metrics. Would be very grateful if you could take a look, thanks!

**Comment by ckkelvinchan** [comment]
> @ckkelvinchan thank you for your comments! Sorry for the late reply I have been busy recently. Indeed this is the reimplementation of the most recent image inpainting paper "Aggregated Contextual Transformations for High-Resolution Image Inpainting". The aim is to integrate AOT-GAN into mmediting.
> 
> I have modified the data augmentation functions so that they only use numpy arrays throughout (please see commit [327bc7f](https://github.com/open-mmlab/mmediting/commit/327bc7f91db9d20c39497e7056af5362ae747ae2) for details).
> 
> I have also pushed the most recent code for the model and evaluation metrics. Would be very grateful if you could take a look, thanks!

Hi @quincylin1 thank you very much for your contribution, and we'd love to integrate AOT-GAN into MMEditing too!

However, it is quite difficult to review the entire implementation in one PR. Do you mind separating them into multiple PRs. For example, shall we do the augmentation (e.g. Crop, Rotation, ...) first? After that, we can work on the model and then the backbone. 

Thanks again!

**Comment by quincylin1** [comment]
> > @ckkelvinchan thank you for your comments! Sorry for the late reply I have been busy recently. Indeed this is the reimplementation of the most recent image inpainting paper "Aggregated Contextual Transformations for High-Resolution Image Inpainting". The aim is to integrate AOT-GAN into mmediting.
> > I have modified the data augmentation functions so that they only use numpy arrays throughout (please see commit [327bc7f](https://github.com/open-mmlab/mmediting/commit/327bc7f91db9d20c39497e7056af5362ae747ae2) for details).
> > I have also pushed the most recent code for the model and evaluation metrics. Would be very grateful if you could take a look, thanks!
> 
> Hi @quincylin1 thank you very much for your contribution, and we'd love to integrate AOT-GAN into MMEditing too!
> 
> However, it is quite difficult to review the entire implementation in one PR. Do you mind separating them into multiple PRs. For example, shall we do the augmentation (e.g. Crop, Rotation, ...) first? After that, we can work on the model and then the backbone.
> 
> Thanks again!

@ckkelvinchan sure. I think the best is to close this pr and then open new ones.

**Comment by ckkelvinchan** [comment]
> > > @ckkelvinchan thank you for your comments! Sorry for the late reply I have been busy recently. Indeed this is the reimplementation of the most recent image inpainting paper "Aggregated Contextual Transformations for High-Resolution Image Inpainting". The aim is to integrate AOT-GAN into mmediting.
> > > I have modified the data augmentation functions so that they only use numpy arrays throughout (please see commit [327bc7f](https://github.com/open-mmlab/mmediting/commit/327bc7f91db9d20c39497e7056af5362ae747ae2) for details).
> > > I have also pushed the most recent code for the model and evaluation metrics. Would be very grateful if you could take a look, thanks!
> > 
> > 
> > Hi @quincylin1 thank you very much for your contribution, and we'd love to integrate AOT-GAN into MMEditing too!
> > However, it is quite difficult to review the entire implementation in one PR. Do you mind separating them into multiple PRs. For example, shall we do the augmentation (e.g. Crop, Rotation, ...) first? After that, we can work on the model and then the backbone.
> > Thanks again!
> 
> @ckkelvinchan sure. I think the best is to close this pr and then open new ones.

@quincylin1 Thank you! Looking forward to working together with you and integrating the model into MMEditing!

**Comment by quincylin1** [comment]
> > > > @ckkelvinchan thank you for your comments! Sorry for the late reply I have been busy recently. Indeed this is the reimplementation of the most recent image inpainting paper "Aggregated Contextual Transformations for High-Resolution Image Inpainting". The aim is to integrate AOT-GAN into mmediting.
> > > > I have modified the data augmentation functions so that they only use numpy arrays throughout (please see commit [327bc7f](https://github.com/open-mmlab/mmediting/commit/327bc7f91db9d20c39497e7056af5362ae747ae2) for details).
> > > > I have also pushed the most recent code for the model and evaluation metrics. Would be very grateful if you could take a look, thanks!
> > > 
> > > 
> > > Hi @quincylin1 thank you very much for your contribution, and we'd love to integrate AOT-GAN into MMEditing too!
> > > However, it is quite difficult to review the entire implementation in one PR. Do you mind separating them into multiple PRs. For example, shall we do the augmentation (e.g. Crop, Rotation, ...) first? After that, we can work on the model and then the backbone.
> > > Thanks again!
> > 
> > 
> > @ckkelvinchan sure. I think the best is to close this pr and then open new ones.
> 
> @quincylin1 Thank you! Looking forward to working together with you and integrating the model into MMEditing!

@ckkelvinchan likewise! I've just created four pr for each respective part of the aot-gan model, please check! I'll close this pr now.
